Лаб. 3 6413 Владимирцев Алексей

Alex8608 opened the pull request:
**ФИО:** Владимирцев Алексей 
**Номер группы:** 6413  
**Номер лабораторной:** 3  
**Номер варианта:** 62 
**Краткое описание предметной области:** Авиакомпания
**Краткое описание добавленных фич:** Поддержка MS SQL Server. DataBase репозитории. Добавление Aspire для орекстрации и запуска сервера и БД. Миграции
Я реализовал сразу 2 и 3 лабораторную работу.

danlla commented on TaxiCompany/TaxiCompany.WebApi/Controllers/CarController.cs:
```diff
@@ -0,0 +1,181 @@
+        var car = await repository.GetAsync(id);
+        if (car == null) return BadRequest($"Car with ID {id} was not found.");
+        var driver = await repositoryDrivers.GetAsync(car.AssignedDriverId);
+        if (driver != null)
+        {
+            driver.AssignedCarId = 0;
+            await repositoryDrivers.PutAsync(driver.Id, driver);
+        }
```
Этот код успешно не выполнится, AssignedCarId это foreign key, он должен соответствовать какому-то объекту
Здесь лучше каскадно удалять и водителя при удалении машины

danlla commented on TaxiCompany/TaxiCompany.WebApi/Controllers/DriverController.cs:
```diff
@@ -0,0 +1,133 @@
+
+        // Получаем список машин с назначенными водителями
+        var assignedDriverIds = (await repositoryCars.GetAsync())
+            .Where(car => car.AssignedDriverId != null)
```
`car.AssignedDriverId` никогда не будет null

danlla commented on TaxiCompany/TaxiCompany.Client/Components/Pages/Cars.razor:
```diff
@@ -0,0 +1,228 @@
+{
+    private List<Car> _cars = new();
+    private List<Driver> _drivers = new();
+    private Car currentCar = new() { Id = 0, Colour = "", Model = "", SerialNumber = "", ReleaseYear = DateTime.Now, AssignedDriverId = 0 };
+    private string carsUrl = "api/car";
+    private string driversUrl = "api/driver";
```
Нарушение в наименовании приватных полей, проверить везде

danlla reviewed (changes requested):
1. Поправить все предупреждения
2. Хотелось бы какого-то разделения в аналитических запросов, но ладно, так тоже пойдет

danlla commented on TaxiCompany/TaxiCompany.Client/Components/Pages/Cars.razor:
```diff
@@ -0,0 +1,229 @@
+{
+    private List<Car> _cars = new();
+    private List<Driver> _drivers = new();
+    private Car currentCar = new() { Id = 0, Colour = "", Model = "", SerialNumber = "", ReleaseYear = DateTime.Now, AssignedDriverId = 0 };
```
`_currentCar`

danlla commented on TaxiCompany/TaxiCompany.Client/Components/Pages/Cars.razor:
```diff
@@ -0,0 +1,229 @@
+    private List<Driver> _drivers = new();
+    private Car currentCar = new() { Id = 0, Colour = "", Model = "", SerialNumber = "", ReleaseYear = DateTime.Now, AssignedDriverId = 0 };
+
+    private string carsUrl = "api/car";
```
`_carsUrl`

danlla commented on TaxiCompany/TaxiCompany.Client/Components/Pages/Cars.razor:
```diff
@@ -0,0 +1,229 @@
+    private Car currentCar = new() { Id = 0, Colour = "", Model = "", SerialNumber = "", ReleaseYear = DateTime.Now, AssignedDriverId = 0 };
+
+    private string carsUrl = "api/car";
+    private string driversUrl = "api/driver";
```
`_driverUrl`

danlla commented on TaxiCompany/TaxiCompany.Client/Components/Pages/Cars.razor:
```diff
@@ -0,0 +1,229 @@
+    private string carsUrl = "api/car";
+    private string driversUrl = "api/driver";
+
+    private bool isAdd = true;
```
`_isAdd`

danlla reviewed (approved):
Тут еще по мелочи остались приватные поля без `_`, ну это не так страшно

Alex8608 committed:
made core entities and unit tests

Alex8608 committed:
Add readme.md, change code of entites ans tests, add dotnet-tests.yml for CI, remove projects that don't need

Alex8608 committed:
edit workflows

Alex8608 committed:
delete constructors and edit mockdata.cs, delete all useless projects, divide fixture class and data class on two different files

Alex8608 committed:
use primary constructor

Alex8608 committed:
add core infrastructure and API layers

Alex8608 committed:
edit readme.md

DmitryKrakhmalev commented on Temp.txt:
Убрать ненужный файл

DmitryKrakhmalev commented on AirlineCompany.Infrastructure/Data/DataSeeder.cs:
```diff
@@ -0,0 +1,217 @@
+    /// <summary>
+    /// Gets seed data for aircraft families
+    /// </summary>
+    public List<AircraftFamily> AircraftFamilies =>
```
Исходя из данных - можно пометить все методы static, как и сам класс.

DmitryKrakhmalev commented on AirlineCompany.Infrastructure/Migrations/20251211101240_InitialCreateWithSeedData.cs:
```diff
@@ -0,0 +1,267 @@
+﻿using System;
```
Добавьте первой строкой `// <auto-generated/>` чтобы анализатор кода не отображал по нему предупреждения.

DmitryKrakhmalev commented on AirlineCompany.Infrastructure/Repositories/AircraftFamilyRepository.cs:
```diff
@@ -0,0 +1,68 @@
+    /// </summary>
+    public int Create(AircraftFamily entity)
+    {
+        dbContext.AircraftFamilies.Add(entity);
```
При работе с EfCore используйте асинхронные методы. Поправьте во всех репозиториях.

DmitryKrakhmalev commented on AirlineCompany.Infrastructure/Repositories/AircraftFamilyRepository.cs:
```diff
@@ -0,0 +1,68 @@
+    public List<AircraftFamily> Read() =>
+         dbContext.AircraftFamilies
+            .AsNoTracking()
+            .ToList();
```
Упростить инициализацию коллекции тут и далее по коду где это возможно

DmitryKrakhmalev commented on AirlineCompany.Application/Services/AircraftFamilyService.cs:
```diff
@@ -0,0 +1,64 @@
+    /// </summary>
+    public AircraftFamilyDTO? GetAircraftFamily(int id)
+    {
+        var entity = repository.Read(id);
```
Аналогично. В сервисах реализуем асинхронные методы.

DmitryKrakhmalev commented on AirlineCompany.Application/Services/AircraftFamilyService.cs:
```diff
@@ -0,0 +1,64 @@
+    /// <summary>
+    /// Converts entity to read DTO
+    /// </summary>
+    private static AircraftFamilyDTO MapReadDto(AircraftFamily entity) =>
```
Мне не очень нравится что в данном классе еще замешана логика по конвертации данных.
Я бы предложил завести какой-нибудь класс-хэлпер который этим будет заниматься. Скорее всего он будет статик, у вас по сути все необходимые данные приходят на вход.
Это бы немного разгрузило эти классы. На своё усмотрение.

DmitryKrakhmalev commented on AirlineCompany.Application/Services/AircraftModelService.cs:
```diff
@@ -0,0 +1,103 @@
+            PassengerCapacity = dto.PassengerCapacity,
+            CargoCapacity = dto.CargoCapacity,
+            AircraftFamilyId = family.Id,
+            //AircraftFamily = family
```
Если этот код не нужен лучше удалить

DmitryKrakhmalev commented on AirlineCompany.Application/Services/AircraftModelService.cs:
```diff
@@ -0,0 +1,103 @@
+        modelRepository.Read()
+            .Where(m => m.AircraftFamilyId == familyId)
+            .Select(MapReadDto)
+            .ToList();
```
Повторюсь что упростить инициализацию коллекций нужно по всему коду, тут тоже.

DmitryKrakhmalev commented on AirlineCompany.Application/Services/AircraftModelService.cs:
```diff
@@ -0,0 +1,103 @@
+    {
+        var family = familyRepository.Read(dto.AircraftFamilyId);
+        if (family == null)
+            throw new ArgumentException("Invalid AircraftFamily ID");
```
```suggestion
        var family = familyRepository.Read(dto.AircraftFamilyId)
            ?? throw new ArgumentException("Invalid AircraftFamily ID");
```

DmitryKrakhmalev commented on AirlineCompany.Application/Services/AnalyticService.cs:
```diff
@@ -0,0 +1,154 @@
+        var flights = flightRepository.Read();
+
+        if (!flights.Any())
+            return new List<FlightDurationDTO>();
```
```suggestion
        if (flights.Count == 0)
            return [];
```

DmitryKrakhmalev commented on AirlineCompany.Application/Services/FlightService.cs:
```diff
@@ -0,0 +1,113 @@
+    /// <summary>
+    /// Converts create DTO to entity
+    /// </summary>
+    private Flight MapDto(FlightCreateDTO dto, AircraftModel model) =>
```
Можно пометить static

DmitryKrakhmalev commented on AirlineCompany.Application/Services/FlightService.cs:
```diff
@@ -0,0 +1,113 @@
+            ArrivalDate = dto.ArrivalDate,
+            Duration = dto.Duration,
+            AircraftModelId = model.Id,
+            //AircraftModel = model
```
Аналогично - если этот код не нужен, лучше его удалите

DmitryKrakhmalev commented on AirlineCompany.AppHost/AppHost.cs:
```diff
@@ -0,0 +1,11 @@
+
+builder.AddProject<Projects.AirlineCompany_API>("airline-api")
+    .WithReference(sql)
+    .WithExternalHttpEndpoints();
```
Добавить ожидание `sql` WaitFor. Ну и в целом имя sql какое то не очень правильное в данном контексте.

Alex8608 replied:
WaitFor не получается применить
CS1660 Не удается преобразовать лямбда-выражение к типу "IResourceBuilder<IResource>", так как он не является типом делегата.
Пытался исправить но не понял как.

DmitryKrakhmalev replied:
Может вы просто не правильно вызываете? Ресурсом у вас выступает db.
```c#
builder.AddProject<Projects.AirlineCompany_Api>("airline-api")
    .WithReference(db)
    .WaitFor(db)
    .WithExternalHttpEndpoints();
```

DmitryKrakhmalev commented on enterprise-development.slnx:
```diff
@@ -0,0 +1,10 @@
+<Solution>
+  <Project Path="AirlineCompany.API/AirlineCompany.API.csproj" />
```
Переименуйте проект в AirlineCompany.Api (PascalCase в именовании). Не забудьте про неймспейсы.

DmitryKrakhmalev commented on AirlineCompany.API/Program.cs:
```diff
@@ -0,0 +1,56 @@
+builder.Services.AddScoped<IRepository<Passenger>, PassengerRepository>();
+builder.Services.AddScoped<IRepository<Ticket>, TicketRepository>();
+
+builder.Services.AddScoped<AircraftFamilyService>();
```
Для сервисов лучше тоже заводить интерфейсы и регистрировать именно через интерфейс. Можно их в проект с DTO положить. 
Если появится у кого то необходимость переопределить логику - он сможет реализовать свой Application слой и спокойно зарегистрировать.

DmitryKrakhmalev commented on AirlineCompany.API/Program.cs:
```diff
@@ -0,0 +1,56 @@
+builder.Services.AddScoped<AnalyticService>();
+
+var connectionString = builder.Configuration.GetConnectionString("DefaultConnection")
+                     ?? throw new InvalidOperationException("Connection string 'DefaultConnection' not found.");
```
Выровняйте лесенку, не очень смотрится. Пусть лучше 4 пробела от предыдущего уровня.

DmitryKrakhmalev commented on AirlineCompany.API/Controllers/AircraftFamiliesController.cs:
```diff
@@ -0,0 +1,88 @@
+    /// </summary>
+    [HttpPost]
+    [ProducesResponseType(201)]
+    public ActionResult Create([FromBody] AircraftFamilyCreateDTO dto)
```
Нужны правки во всех контроллерах:
1. Типизируйте возвращаемый результат. ActionResult является дженериком и может принимать тип возвращаемого значения. ProducesResponseType - аналогично.
https://learn.microsoft.com/ru-ru/aspnet/core/web-api/action-return-types?view=aspnetcore-10.0#actionresultt-type
2. В DTO на создание лучше добавить валидационные атрибуты чтобы все это валидировалось.
https://learn.microsoft.com/ru-ru/aspnet/core/mvc/models/validation?view=aspnetcore-8.0#validation-attributes
3. Добавьте обработку ошибок и логгирование. Ошибку сохраняйте в логи а пользователю сообщайте краткий ответ не раскрывая места возникновения проблемы.

Даже если кажется что в данном методе никакая обработка ошибок не нужна - а что будет если у вас база данных станет не доступна или еще что произойдет? Лучше это все учитывать.

DmitryKrakhmalev commented on AirlineCompany.API/Controllers/AircraftFamiliesController.cs:
```diff
@@ -0,0 +1,88 @@
+    /// <summary>
+    /// Update aircraft family by ID
+    /// </summary>
+    [HttpPut("{id}")]
```
Можно сразу же и проверку по типу добавить.
```suggestion
    [HttpPut("{id:int}")]
```

DmitryKrakhmalev reviewed (changes requested):
Есть замечания которые нужно устранить.

Alex8608 committed:
complete API implementation
- Add validation and async support
- Implement error handling
- Fix migration annotations

Alex8608 committed:
change name of projects and namespaces (PascalCase)

Alex8608 committed:
edit CI

Alex8608 committed:
edit CI again

Alex8608 committed:
rename API to Api and DTO to Dto (force Git to recognize rename)

DmitryKrakhmalev commented on AirlineCompany.Application/Services/AnalyticService.cs:
```diff
@@ -0,0 +1,156 @@
+    /// <summary>
+    /// Helper method to convert Flight entity to FlightDTO
+    /// </summary>
+    private FlightDTO MapToFlightDTO(Flight entity)
```
Именуем в PascalCase, даже акронимы. Переименуйте в FlightDto а метод в MapToFlightDto. Аналогично в других местах. Также можно пометить static.

DmitryKrakhmalev commented on AirlineCompany.Application/Services/TicketService.cs:
```diff
@@ -0,0 +1,145 @@
+                entity.Flight.AircraftModel.Range,
+                entity.Flight.AircraftModel.PassengerCapacity,
+                entity.Flight.AircraftModel.CargoCapacity,
+                new AircraftFamilyDTO(
```
Я предлагал вам завести отдельный класс для конвертации данных, это бы сняло нагрузку с служб, а также можно было бы использовать эти методы в любом месте. Сейчас код порастает какой то матрешкой по конвертации и в разных классах имеется схожий код по созданию объектов путем копирования данных. 
Я бы все таки рекомендовал выделить данные методы в какой то дополнительный хэлпер класс

DmitryKrakhmalev reviewed (changes requested):
Небольшие правки

Alex8608 committed:
rename files and methods DTO to Dto (PascalCase)

Alex8608 committed:
Add MapperHelper and edit services using MapperHelper
